clab/clab.go

```diff
@@ -520,29 +534,123 @@ func (c *CLab) scheduleNodes(ctx context.Context, maxWorkers int,
 					log.Errorf("failed pre-deploy phase for node %q: %v", node.Config().ShortName, err)
 					continue
 				}
+
+				// enter the create phase
+				err = dm.EnterPhase(node.Config().ShortName, types.WaitForCreate)
+				if err != nil {
+					log.Error(err)
+				}
+
 				// Deploy
 				err = node.Deploy(ctx, &nodes.DeployParams{})
 				if err != nil {
 					log.Errorf("failed deploy phase for node %q: %v", node.Config().ShortName, err)
 					continue
 				}
 
+				// we need to update the node's state with runtime info (e.g. the mgmt net ip addresses)
+				// before continuing with the post-deploy phase (for e.g. certificate creation)
+				err = node.UpdateConfigWithRuntimeInfo(ctx)
+				if err != nil {
+					log.Errorf("failed to update node runtime information for node %s: %v", node.Config().ShortName, err)
+				}
+
+				dm.SignalDone(node.GetShortName(), types.WaitForCreate)
+
+				err = dm.EnterPhase(node.GetShortName(), types.WaitForCreateLinks)
+				if err != nil {
+					log.Error(err)
+				}
+
 				err = node.DeployLinks(ctx)
 				if err != nil {
 					log.Errorf("failed deploy links for node %q: %v", node.Config().ShortName, err)
 					continue
 				}
 
-				// signal to dependency manager that this node is done with creation
-				dm.SignalDone(node.Config().ShortName, dependency_manager.NodeStateCreated)
+				// DeployLinks does not necessarily create all the links already.
+				// veth links might be created by the peer node, if it was not already
+				// deployed properly. Hence we need to wait for all the links to be created.
+				// we just wait if there is actually a dependency on this state, otherwise
+				// we head on.
+				count, err := dm.GetDependerCount(node.GetShortName(), types.WaitForCreateLinks)
+				if err != nil {
+					log.Error(err)
+				}
```

Comment:
@steiler I am not sure this GetDependerCount logic is required at all.
Don't we always want to ensure that all links were created?

All other stages require links to be provisioned, so maybe we can just drop this additional logic?

Reply:
we'll the thing is, that the first node, that is deployed, cannot create any links, because all the other nodes are not there yet. Waiting for all the nodes to complete to then create the links is something we got rid of lately. All the links these days are created by the second node (per link). If we need all the links to be created, we have to explicitly wait for all the links. If there is no dependency on the creation of links, why would we wait, we can simply carry on and speed up the deployment.
What I did was to change the count to a simple boolean. Because I realized the count does not really matter, it is more the "if" (we have to wait) that matters. 
This logic also reapeats itself on the healthy and the exit phase ...
This is crutial, since many containers do not have a healthcheck defined, so we would never be done with the waiting for healthy. So clan deploy would never terminate. Same applies to the exit state. Mostly you will not wait for exit state, which is basically our mechanism that enables init containers. But for all other containers we don't want to wait for the exit state, because clab would "never" terminate.